File: backend/src/main/java/com/langleon/dsobuildsim/items/dto/ItemInstanceDTO.java
package com.langleon.dsobuildsim.items.dto;

import com.langleon.dsobuildsim.common.StatType;
import com.langleon.dsobuildsim.enchantments.dto.EnchantmentDTO;
import com.langleon.dsobuildsim.gems.dto.GemInstanceDTO;
import com.langleon.dsobuildsim.items.core.enums.ItemCategory;

import java.util.List;
import java.util.Map;

public record ItemInstanceDTO(
        ItemCategory itemCategory,
        String itemType,
        int level,
        Map<StatType, Double> baseValues,
        List<GemInstanceDTO> gems,
        List<EnchantmentDTO> enchantments,

        //optional fields depending on category
        Map<StatType, Double> uniqueBaseValues,
        List<EnchantmentDTO> uniqueEnchantments
) {
}

File: frontend/src/app/models/instanceDTOs/ItemInstanceDTO.ts
import { StatType } from "../../enums/StatType";
import { EnchantmentDTO } from "../gamedataDTOs/EnchantmentDTO";
import { GemInstanceDTO } from "./GemInstanceDTO";

export interface ItemInstanceDTO {
    itemCategory: string;
    itemType: string;
    level: number;
    baseValues: Record<StatType, number>;
    gems: GemInstanceDTO[];
    enchantments: EnchantmentDTO[];

    //optional fields depending on item category
    uniqueBaseValues: Record<string, number>;
    uniqueEnchantments: EnchantmentDTO[];
}